ghcide/src/Development/IDE/Core/Compile.hs

```diff
@@ -269,40 +283,69 @@ captureSplicesAndDeps TypecheckHelpers{..} env k = do
 
             -- Find the linkables for the modules we need
-           ; let needed_mods = mkUniqSet [ moduleName mod
+           ; let needed_mods = mkUniqSet [
+#if MIN_VERSION_ghc(9,3,0)
+                                           mod -- We need the whole module for 9.4 because of multiple home units modules may have different unit ids
+#else
+                                           moduleName mod -- On <= 9.2, just the name is enough because all unit ids will be the same
+#endif
+
 #if MIN_VERSION_ghc(9,2,0)
                                          | n <- concatMap (uniqDSetToList . bcoFreeNames) $ bc_bcos bcos
 #else
                                          | n <- uniqDSetToList (bcoFreeNames bcos)
 #endif
                                          , Just mod <- [nameModule_maybe n] -- Names from other modules
                                          , not (isWiredInName n) -- Exclude wired-in names
-                                         , moduleUnitId mod == uid -- Only care about stuff from the home package
+                                         , moduleUnitId mod `elem` home_unit_ids -- Only care about stuff from the home package set
                                          ]
-                 hpt = hsc_HPT hsc_env
-                 uid = homeUnitId_ dflags
-                 mods_transitive = getTransitiveMods hpt needed_mods
-                  -- Non det OK as we will put it into maps later anyway
-                 mods_transitive_list = nonDetEltsUniqSet mods_transitive
-
-           ; lbs <- getLinkables [toNormalizedFilePath' file | mod <- mkHomeModule
+                 home_unit_ids =
+#if MIN_VERSION_ghc(9,3,0)
+                    map fst (hugElts $ hsc_HUG hsc_env)
+#else
+                    [homeUnitId_ dflags]
+#endif
+                 mods_transitive = getTransitiveMods hsc_env needed_mods
+                 -- If we don't support multiple home units, ModuleNames are sufficient because all the units will be the same
+                 mods_transitive_list = 
+#if MIN_VERSION_ghc(9,3,0)
+                                         mapMaybe nodeKeyToInstalledModule $ Set.toList mods_transitive
+#else
+                                         map 
 #if MIN_VERSION_ghc(9,0,0)
-                                                                      (hscHomeUnit hsc_env)
+                                            (mkModule (homeUnitId_ dflags))
+#else
+                                            (InstalledModule (toInstalledUnitId $ homeUnitId_ dflags))
+#endif
+                                        -- Non det OK as we will put it into maps later anyway
+                                       $ nonDetEltsUniqSet mods_transitive
+#endif
+
+#if MIN_VERSION_ghc(9,3,0)
+           ; moduleLocs <- readIORef (fcModuleCache $ hsc_FC hsc_env)
 #else
-                                                                      uid
+           ; moduleLocs <- readIORef (hsc_FC hsc_env)
 #endif
```

Comment:
This sets a new record.
Would it not be just better to have 3 versions of this code, one for each GHC api?

Reply:
I've reduced the nesting, is that any better?